File: src/main/java/io/vproxy/vfx/util/algebradata/XYZTData.java
package io.vproxy.vfx.util.algebradata;

public class XYZTData implements AlgebraData<XYZTData> {
    public final double x;
    public final double y;
    public final double z;
    public final double t;

    public XYZTData(double x, double y, double z, double t) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.t = t;
    }

    @Override
    public XYZTData plus(XYZTData other) {
        return new XYZTData(x + other.x, y + other.y, z + other.z, t + other.t);
    }

    @Override
    public XYZTData minus(XYZTData other) {
        return new XYZTData(x - other.x, y - other.y, z - other.z, t - other.t);
    }

    @Override
    public XYZTData multiply(double v) {
        return new XYZTData(x * v, y * v, z * v, t * v);
    }

    @Override
    public XYZTData dividedBy(double v) {
        return new XYZTData(x / v, y / v, z / v, t / v);
    }

    @Override
    public String toString() {
        return "XYZTData(" + x + ", " + y + ", " + z + ", " + t + ")";
    }
}

File: src-ts/src/util/algebradata/XYZTData.ts
import { AlgebraData } from './AlgebraData.js';

export class XYZTData implements AlgebraData<XYZTData> {
  constructor(
    public readonly x: number,
    public readonly y: number,
    public readonly z: number,
    public readonly t: number,
  ) {}

  plus(o: XYZTData): XYZTData {
    return new XYZTData(this.x + o.x, this.y + o.y, this.z + o.z, this.t + o.t);
  }
  minus(o: XYZTData): XYZTData {
    return new XYZTData(this.x - o.x, this.y - o.y, this.z - o.z, this.t - o.t);
  }
  multiply(v: number): XYZTData {
    return new XYZTData(this.x * v, this.y * v, this.z * v, this.t * v);
  }
  dividedBy(v: number): XYZTData {
    return new XYZTData(this.x / v, this.y / v, this.z / v, this.t / v);
  }

  toString(): string { return `XYZTData(${this.x}, ${this.y}, ${this.z}, ${this.t})`; }
}
